xonsh/procs/specs.py

```diff
@@ -707,27 +708,41 @@ def resolve_redirects(self):
     def resolve_alias(self):
         """Sets alias in command, if applicable."""
         cmd0 = self.cmd[0]
-        spec_modifiers = []
         if cmd0 in self.alias_stack:
             # Disabling the alias resolving to prevent infinite loop in call stack
             # and futher using binary_loc to resolve the alias name.
             self.alias = None
             return
 
         if callable(cmd0):
-            alias = cmd0
+            self.alias = cmd0
         else:
+            found_spec_modifiers = []
+            found_return_command = []
             if isinstance(XSH.aliases, dict):
                 # Windows tests
                 alias = XSH.aliases.get(cmd0, None)
+                if alias is not None:
+                    alias = alias + self.cmd[1:]
             else:
-                alias = XSH.aliases.get(cmd0, None, spec_modifiers=spec_modifiers)
+                alias = XSH.aliases.get(
+                    self.cmd,
+                    None,
+                    spec_modifiers=found_spec_modifiers,
+                    found_return_command=found_return_command,
+                )
             if alias is not None:
                 self.alias_name = cmd0
-        self.alias = alias
-        if spec_modifiers:
-            for mod in spec_modifiers:
-                self.add_spec_modifier(mod)
+                if callable(alias[0]):
+                    self.alias = alias[0]
+                    self.cmd = [cmd0] + alias[1:]
+                else:
+                    self.alias = alias
+
+            self.alias_return_command = bool(found_return_command)
+            if found_spec_modifiers:
+                for mod in found_spec_modifiers:
+                    self.add_spec_modifier(mod)
```

Comment:
This change increases the cyclomatic complexity of this function from 7 to 9, where 10 is often considered a good upper limit to avoid.

I don't have the mental bandwidth right now to decipher this logic, and maybe this approach is as simple as it can be, but if you think it might be possible to separate some concerns, either with comments that describe the thinking, or better with functions/methods that encapsulate and document smaller steps, that would help me (and likely others) understand what's going on here.